File: src/main/java/com/simprints/libsimprints/Registration.java
package com.simprints.libsimprints;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import androidx.annotation.NonNull;

@SuppressWarnings({"unused", "WeakerAccess"})
public class Registration implements Parcelable {

    private String guid;
    private Map<FingerIdentifier, byte[]> templates;

    /**
     * This constructor creates a new registration
     *
     * @param guid Global user id
     */
    public Registration(String guid) {
        this.guid = guid;
        templates = new HashMap<>();
    }

    protected Registration(Parcel in) {
        this.guid = in.readString();
        int nbOfTemplates = in.readInt();
        this.templates = new HashMap<>(nbOfTemplates);
        for (int i = 0; i < nbOfTemplates; i++) {
            FingerIdentifier fingerId = FingerIdentifier.values()[in.readInt()];
            byte[] fingerTemplate = new byte[in.readInt()];
            in.readByteArray(fingerTemplate);
            this.templates.put(fingerId, fingerTemplate);
        }
    }

    public static final Creator<Registration> CREATOR = new Creator<Registration>() {
        @Override
        public Registration createFromParcel(Parcel in) {
            return new Registration(in);
        }

        @Override
        public Registration[] newArray(int size) {
            return new Registration[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(guid);
        dest.writeInt(templates.size());
        for (Map.Entry<FingerIdentifier, byte[]> e : templates.entrySet()) {
            dest.writeInt(e.getKey().ordinal());
            dest.writeInt(e.getValue().length);
            dest.writeByteArray(e.getValue());
        }
    }

    public String getGuid() {
        return guid;
    }

    public void setTemplate(FingerIdentifier fingerId, @NonNull byte[] fingerTemplate) {
        this.templates.put(fingerId, fingerTemplate);
    }

    public byte[] getTemplate(FingerIdentifier fingerId) {
        if (templates.containsKey(fingerId)) {
            return templates.get(fingerId);
        } else {
            return new byte[0];
        }
    }

    @Override
    public boolean equals(Object o) {
        if (o == this) {
            return true;
        }
        if (!(o instanceof Registration)) {
            return false;
        }
        Registration other = (Registration) o;
        if (!guid.equals(other.guid)) {
            return false;
        }
        for (FingerIdentifier fingerId : FingerIdentifier.values()) {
            if (!Arrays.equals(templates.get(fingerId), other.templates.get(fingerId))) {
                return false;
            }
        }
        return true;
    }

    @Override
    public int hashCode() {
        int hash = guid.hashCode();
        for (FingerIdentifier fingerId : FingerIdentifier.values()) {
            hash = hash * 17 + Arrays.hashCode(templates.get(fingerId));
        }
        return hash;
    }
}

File: src/main/java/com/simprints/libsimprints/Registration.kt
package com.simprints.libsimprints

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator

@Deprecated("Use contracts.data.Enrolment instead")
data class Registration @JvmOverloads constructor(
    val guid: String,
    private val templates: MutableMap<FingerIdentifier, ByteArray> = mutableMapOf(),
) : Parcelable {
    fun setTemplate(
        fingerId: FingerIdentifier,
        fingerTemplate: ByteArray,
    ) {
        templates[fingerId] = fingerTemplate
    }

    fun getTemplate(fingerId: FingerIdentifier): ByteArray = templates[fingerId] ?: byteArrayOf()

    override fun describeContents(): Int = 0

    override fun writeToParcel(
        dest: Parcel,
        flags: Int,
    ) {
        dest.writeString(guid)
        dest.writeInt(templates.size)

        for ((fingerId, fingerTemplate) in templates) {
            dest.writeInt(fingerId.ordinal)
            dest.writeInt(fingerTemplate.size)
            dest.writeByteArray(fingerTemplate)
        }
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + guid.hashCode()
        result = 31 * result + templates.hashCode()
        return result
    }

    override fun equals(other: Any?): Boolean {
        if (other === this) return true
        if (other !is Registration) return false
        if (guid != other.guid) return false
        for (fingerId in FingerIdentifier.entries) {
            if (!templates[fingerId].contentEquals(other.templates[fingerId])) return false
        }
        return true
    }

    companion object {
        @JvmField
        val CREATOR = object : Creator<Registration> {
            override fun createFromParcel(parcel: Parcel): Registration? {
                val guid = parcel.readString().orEmpty()
                val nbOfTemplates = parcel.readInt()

                val templates = mutableMapOf<FingerIdentifier, ByteArray>()
                for (i in 0 until nbOfTemplates) {
                    val fingerId = FingerIdentifier.entries[parcel.readInt()]
                    val fingerTemplate = ByteArray(parcel.readInt())
                    parcel.readByteArray(fingerTemplate)
                    templates[fingerId] = fingerTemplate
                }
                return Registration(guid, templates)
            }

            override fun newArray(p0: Int): Array<out Registration?> = arrayOfNulls(p0)
        }
    }
}
